### PR TITLE
Add PrivateRecursive type annotations

### PR DESCRIPTION
**Pull request by jdalton**
### What does this PR do?

~~Add alwaysInline and PrivateRecursive annotations~~

Add PrivateRecursive type annotations

-->

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [ ] Code changes

### CODE REVIEW

File: src/js/builtins/Module.ts [hunk 1]
```diff
@@ -3,14 +3,16 @@ export function main() {
   return $requireMap.$get(Bun.main);
 }
 
-$visibility = "Private";
+$visibility = "PrivateRecursive";
+$alwaysInline;
```
Comment (jdalton) [inline src/js/builtins/Module.ts]: Adds `@alwaysInline` Like JSC does for `Array#filter`:
https://github.com/WebKit/WebKit/commit/6cb56ee3051f43ee8030b9c34b4ea035a2effa6a

File: src/js/builtins/ImportMetaObject.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 type ImportMetaObject = Partial<ImportMeta>;
 
-$visibility = "Private";
+$visibility = "PrivateRecursive";
 export function loadCJS2ESM(this: ImportMetaObject, resolvedSpecifier: string) {
```
Comment (jdalton) [inline src/js/builtins/ImportMetaObject.ts]: Protects call stack of internals recursively. (see https://github.com/WebKit/WebKit/commit/37f814d56d6446ea011af18ee65b6644c1127b4e).

**Comment by jdalton** [comment]
This PR is mainly to introduce these, as I found them cool. Y'all may have a better idea of where to add them and see if it makes a difference.

File: src/js/builtins/ImportMetaObject.ts [hunk 3]
```diff
@@ -109,7 +109,7 @@ export function requireESM(this: ImportMetaObject, resolved) {
   return exports;
 }
 
-$visibility = "Private";
+$visibility = "PrivateRecursive";
 export function internalRequire(this: ImportMetaObject, id) {
   var cached = $requireMap.$get(id);
```
Comment (jdalton) [inline src/js/builtins/ImportMetaObject.ts]: Depending on how much these restrict the stack we may not want to do this. It's mostly trying them on.

File: src/js/builtins/Module.ts (hunk 1 above)
Comment (paperclover) [inline src/js/builtins/Module.ts]: i wonder if this actually does anything. this function is almost always `.bind()`ed to the module record.

File: src/js/builtins/Module.ts [hunk 4]
```diff
@@ -3,14 +3,16 @@ export function main() {
...
 export function require(this: CommonJSModuleRecord, id: string) {
   return $tailCallForwardArguments($overridableRequire, this);
 }
 
 // overridableRequire can be overridden by setting `Module.prototype.require`
 $overriddenName = "require";
-$visibility = "Private";
+$visibility = "PrivateRecursive";
+$alwaysInline;
```
Comment (paperclover) [inline src/js/builtins/Module.ts]: similarly on this one, i wonder if this actually ever gets inlined, considering it is a part of the `tailCallForwardArguments`.

**Review by paperclover — COMMENTED** [review]
i dont think making any of these functions PrivateRecursive is going to have any specific effect. but i dont think it is of harm to do it either.

so long as stack traces of a required module dont get hidden (`require()` calls into the user code, so perhaps that gets recursively marked), then it's okay.

but i'm not really sold on this change needing to happen at all. adding the typedef is good tho, we should add that at the very minimum.

File: src/js/builtins/Module.ts (hunk 1 above)
Comment (jdalton) [reply src/js/builtins/Module.ts]: Ya, this PR was mainly a "see something, say something" moment. I think knowing we have the ability to specify `$awaysInline` is powerful. Not likely the place for it though you are right.

File: src/js/builtins/Module.ts (hunk 4 above)
Comment (jdalton) [reply src/js/builtins/Module.ts]: My concern is that this would hide things we DON'T want hidden too.

**Comment by jdalton** [comment]
@paperdave 
> but i'm not really sold on this change needing to happen at all. adding the typedef is good tho, we should add that at the very minimum.

Same. I noticed in WebKit's code they reach for `PrivateRecursive` now more than `Private` but our usage/milage may vary from WebKit core.

**Comment by jdalton** [comment]
Updated to just the d.ts type 😸

**Pull request merged**